packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,116 @@
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, CfnConstraintResource>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct, options: TagUpdatesOptions) {
+    const associationKey = this.associationPrecheck(scope, portfolio, product);
+    InputValidator.validateLength(this.prettyPrintAssociation(scope, portfolio, product), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey)!.has(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptedMessageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? Allowed.NOT_ALLOWED : Allowed.ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(scope, portfolio, product)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value, scope.stack.node.addr);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const cfnTagOption = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, cfnTagOption);
+
+          new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+            resourceId: resourceId,
+            tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
```

Comment:
Unless I'm missing something, `this.tagOptionMap.get(tagOptionKey)!` is literally `cfnTagOption` (it's set on the previous line).

Reply:
punting new PR.  also for future reference this was accidentally introduced when thought was making the calls idempotent.  Because a tag option can be associated with multiple resources (hence separate constructs) so we maintained the map if we had fetched an existing tagOption for a new association.